### PR TITLE
refactor delete and getMetadata,

### PR DESCRIPTION
**Pull request by JacquesBorsenberger**
* move services in their own package

Signed-off-by: jacques borsenberger <jacques.borsenberger@rte-france.com>

### CODE REVIEW

File: src/main/java/org/gridsuite/explore/server/services/ExploreService.java [hunk 1]
```diff
@@ -143,42 +151,35 @@ public ExploreService(
...
-            } else if (elementAttributes.getType().equals(CONTINGENCY_LIST)) {
-                contingencyListService.deleteContingencyList(elementAttributes.getElementUuid()).subscribe();
-            } else if (elementAttributes.getType().equals(FILTER)) {
-                filterService.deleteFilter(elementAttributes.getElementUuid()).subscribe();
+            IDirectoryElementsService service = genericServices.get(elementAttributes.getType());
+            if (service == null) {
+                return Mono.error(new ExploreException(UNKNOWN_ELEMENT_TYPE, "Unknown element type :" + elementAttributes.getType()));
             }
-            return directoryService.deleteElement(id, userId);
+            return service.delete(id, userId).doOnSuccess(e -> directoryService.deleteElement(id, userId));
```
Comment (etiennehomer) [inline src/main/java/org/gridsuite/explore/server/services/ExploreService.java:158]: return service.delete(id, userId).doOnSuccess(e -> directoryService.deleteElement(id, userId).**subscribe()**); ?
The directory delete isn't executed

**Review by etiennehomer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/explore/server/services/ExploreService.java (hunk 1 above)
Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/explore/server/services/ExploreService.java:158]: a subscribe is missing...

Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/explore/server/services/ExploreService.java]: done

File: src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java [hunk 2]
```diff
@@ -49,15 +52,16 @@ public void setActionsServerBaseUri(String actionsServerBaseUri) {
...
 
-    public Mono<Void> deleteContingencyList(UUID id) {
+    public Mono<Void> delete(UUID id, String userId) {
         String path = UriComponentsBuilder.fromPath(DELIMITER + ACTIONS_API_VERSION + "/contingency-lists/{id}")
             .buildAndExpand(id)
             .toUriString();
 
         return webClient.delete()
             .uri(actionsServerBaseUri + path)
+            .header(HEADER_USER_ID, userId)
```
Comment (etiennehomer) [inline src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java:62]: userId isn't used in actions-server on this endpoint. 
You added it cause you think it gonna be used later ?

File: src/main/java/org/gridsuite/explore/server/services/FilterService.java [hunk 3]
```diff
@@ -80,15 +82,16 @@ public void setFilterServerBaseUri(String filterServerBaseUri) {
...
 
-    public Mono<Void> deleteFilter(UUID id) {
+    public Mono<Void> delete(UUID id, String userId) {
         String path = UriComponentsBuilder.fromPath(DELIMITER + FILTER_SERVER_API_VERSION + "/filters/{id}")
                 .buildAndExpand(id)
                 .toUriString();
 
         return webClient.delete()
                 .uri(filterServerBaseUri + path)
+                .header(HEADER_USER_ID, userId)
```
Comment (etiennehomer) [inline src/main/java/org/gridsuite/explore/server/services/FilterService.java:92]: Same question

Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/explore/server/services/FilterService.java:92]: to have the same interface everywhere, and yes possibly use it (trace)

File: src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java (hunk 2 above)
Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java:62]: to have the same interface everywhere, and yes possibly use it (trace)

Comment (etiennehomer) [reply src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java:62]: Ok

**Review by etiennehomer — DISMISSED** [review]

File: src/main/java/org/gridsuite/explore/server/services/ExploreService.java [hunk 4]
```diff
@@ -143,42 +152,36 @@ public ExploreService(
...
             }
-            return directoryService.deleteElement(id, userId);
+            return service.delete(id, userId).doOnSuccess(e -> directoryService.deleteElement(id, userId).subscribe());
         });
     }
 
-    public Mono<List<ElementAttributes>> getElementsMetadata(List<UUID> ids) {
+    public Flux<ElementAttributes> getElementsMetadata(List<UUID> ids) {
         Mono<Map<UUID, ElementAttributes>> elementsAttributesMono = directoryService.getElementsAttribute(ids).collect(Collectors.toMap(ElementAttributes::getElementUuid, Function.identity()));
+        return elementsAttributesMono.flatMapMany(elementsAttributes -> {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/explore/server/services/ExploreService.java:165]: I think it is too complex to mix stream et flux
why we do not use `GroupedFlux` and use only flux ?
`        Flux<GroupedFlux<String, ElementAttributes>> elementsAttributesFlux = directoryService.getElementsAttribute(ids).groupBy(e -> e.getType());
`

File: src/main/java/org/gridsuite/explore/server/services/ExploreService.java [hunk 5]
```diff
@@ -143,42 +152,36 @@ public ExploreService(
...
         return directoryService.getElementInfos(id).flatMap(elementAttributes -> {
-            if (elementAttributes.getType().equals(STUDY)) {
-                studyService.deleteFromStudyServer(elementAttributes.getElementUuid(), userId).subscribe();
-            } else if (elementAttributes.getType().equals(CONTINGENCY_LIST)) {
-                contingencyListService.deleteContingencyList(elementAttributes.getElementUuid()).subscribe();
-            } else if (elementAttributes.getType().equals(FILTER)) {
-                filterService.deleteFilter(elementAttributes.getElementUuid()).subscribe();
+            IDirectoryElementsService service = genericServices.get(elementAttributes.getType());
+            if (service == null) {
+                return Mono.error(new ExploreException(UNKNOWN_ELEMENT_TYPE, "Unknown element type :" + elementAttributes.getType()));
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/explore/server/services/ExploreService.java:157]: Perhaps a function which return a mono like this :
`    private Mono<IDirectoryElementsService> getGenericServices(String elementType) {
        return genericServices.get(elementType) != null ?  Mono.just(genericServices.get(elementType)) : Mono.error(createEquipmentTypeUnknown(elementType));
    }`

**Review by SlimaneAmar — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/explore/server/ExploreException.java [hunk 6]
```diff
@@ -11,23 +11,24 @@
...
-class ExploreException extends RuntimeException {
+public class ExploreException extends RuntimeException {
 
     public enum Type {
         STUDY_NOT_FOUND,
         CONTINGENCY_LIST_NOT_FOUND,
         FILTER_NOT_FOUND,
-        NOT_ALLOWED;
+        NOT_ALLOWED,
+        UNKNOWN_ELEMENT_TYPE;
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/explore/server/ExploreException.java:21]: Need to add in the RestResponseEntityExceptionHandler et unit test ?

File: src/main/java/org/gridsuite/explore/server/services/ExploreService.java [hunk 7]
```diff
@@ -143,42 +152,36 @@ public ExploreService(
...
-            Mono<List<Map<String, Object>>> contingencyListMetadataMono = contingencyListService.getContingencyListMetadata(contingencyListsUuids).collectList();
-
-            Mono<Tuple2<List<Map<String, Object>>, List<Map<String, Object>>>> metadata = Mono.zip(filtersMetadataMono, contingencyListMetadataMono);
-
-            return metadata.map(data -> {
-                Map<String, Map<String, Object>> filtersMetadataMap = data.getT1().stream().collect(Collectors.toMap(e -> e.get("id").toString(), Function.identity()));
-                Map<String, Map<String, Object>> contingenciesMetadataMap = data.getT2().stream().collect(Collectors.toMap(e -> e.get("id").toString(), Function.identity()));
+            if (!genericServices.keySet().containsAll(uuidsByType.keySet())) {
+                return Flux.error(() -> new ExploreException(UNKNOWN_ELEMENT_TYPE, "Unknown element type "));
+            }
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/explore/server/services/ExploreService.java:171]: Move  the unknow type exception creation in ExploreException class ?
`    public static ExploreException createEquipmentTypeUnknown(String type) {
        Objects.requireNonNull(type);
        return new ExploreException(Type.UNKNOWN_ELEMENT_TYPE, "The element type : " + type + " is unknown");
    }`

File: src/main/java/org/gridsuite/explore/server/services/FilterService.java [hunk 8]
```diff
@@ -60,7 +62,7 @@ public void setFilterServerBaseUri(String filterServerBaseUri) {
         return webClient.put()
                 .uri(filterServerBaseUri + path)
                 .retrieve()
-                .onStatus(httpStatus -> httpStatus == HttpStatus.NOT_FOUND, clientResponse -> Mono.error(new ExploreException(FILTER_NOT_FOUND)))
+                .onStatus(httpStatus -> httpStatus == HttpStatus.NOT_FOUND, ClientResponse::createException)
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/explore/server/services/FilterService.java:65]: Why ?

File: src/main/java/org/gridsuite/explore/server/services/FilterService.java [hunk 9]
```diff
@@ -80,15 +82,16 @@ public void setFilterServerBaseUri(String filterServerBaseUri) {
...
         String path = UriComponentsBuilder.fromPath(DELIMITER + FILTER_SERVER_API_VERSION + "/filters/{id}")
                 .buildAndExpand(id)
                 .toUriString();
 
         return webClient.delete()
                 .uri(filterServerBaseUri + path)
+                .header(HEADER_USER_ID, userId)
                 .retrieve()
-                .onStatus(httpStatus -> httpStatus != HttpStatus.OK, r -> Mono.empty())
+                .onStatus(httpStatus -> httpStatus != HttpStatus.OK, r -> Mono.error(() -> new ExploreException(ExploreException.Type.NOT_ALLOWED)))
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/explore/server/services/FilterService.java:94]: Why is it different than the not found above ?

File: src/main/java/org/gridsuite/explore/server/services/ServicesConstants.java [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * @author Jacques Borsenberger <jacques.borsenberger at rte-france.com>
+ */
+
+public final class ServicesConstants {
+    private ServicesConstants() { }
+
+    static final String HEADER_USER_ID = "userId";
+
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/explore/server/services/ServicesConstants.java:17]: Move into IDirectoryElementsService class ?

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java [hunk 11]
```diff
@@ -168,6 +172,15 @@ public MockResponse dispatch(RecordedRequest request) throws InterruptedExceptio
                     return new MockResponse().setResponseCode(200);
                 } else if (path.matches("/v1/filters/.*/replace-with-script") && "PUT".equals(request.getMethod())) {
                     return new MockResponse().setResponseCode(200);
+                } else if ("DELETE".equals(request.getMethod())) {
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/explore/server/ExploreTest.java:175]: Why we do not mock the bean service instead ?

**Review by SlimaneAmar — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/explore/server/services/ExploreService.java (hunk 5 above)
Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/explore/server/services/ExploreService.java]: done

File: src/main/java/org/gridsuite/explore/server/services/ExploreService.java (hunk 4 above)
Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/explore/server/services/ExploreService.java]: done

File: src/main/java/org/gridsuite/explore/server/services/ServicesConstants.java (hunk 10 above)
Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/explore/server/services/ServicesConstants.java]: done

File: src/main/java/org/gridsuite/explore/server/services/ExploreService.java (hunk 7 above)
Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/explore/server/services/ExploreService.java]: done

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java (hunk 11 above)
Comment (JacquesBorsenberger) [reply src/test/java/org/gridsuite/explore/server/ExploreTest.java:177]: need to test complete metadata

File: src/main/java/org/gridsuite/explore/server/services/FilterService.java (hunk 9 above)
Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/explore/server/services/FilterService.java]: hramonized

**Review by SlimaneAmar — APPROVED** [review]

File: src/main/java/org/gridsuite/explore/server/services/IDirectoryElementsService.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * @author Jacques Borsenberger <jacques.borsenberger at rte-france.com>
+ */
+
+interface IDirectoryElementsService {
+    String HEADER_USER_ID = "userId";
+
+    default Flux<Map<String, Object>> getMetadata(List<UUID> uuidList) {
```
[the remark this answers is not in the document]
Comment (etiennehomer) [reply src/main/java/org/gridsuite/explore/server/services/IDirectoryElementsService.java:26]: I think there's never a S to data

**Review by etiennehomer — APPROVED** [review]
Test ok

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java [hunk 13]
```diff
@@ -311,20 +334,21 @@ public void testReplaceFilterWithScript() {
...
 
     @Test
     public void testDeleteElement() {
-        deleteElement(FILTER_UUID);
-        deleteElement(PRIVATE_STUDY_UUID);
-        deleteElement(CONTINGENCY_LIST_UUID);
+        deleteElement(FILTER_UUID, OK);
+        deleteElement(PRIVATE_STUDY_UUID, OK);
+        deleteElement(CONTINGENCY_LIST_UUID, OK);
+        deleteElement(INVALID_ELEMENT_UUID, INTERNAL_SERVER_ERROR);
```
[the remark this answers is not in the document]
Comment (JacquesBorsenberger) [reply src/test/java/org/gridsuite/explore/server/ExploreTest.java]: done

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java [hunk 14]
```diff
@@ -311,20 +334,21 @@ public void testReplaceFilterWithScript() {
                 .expectStatus().isOk();
     }
 
-    public void deleteElement(UUID elementUUid) {
+    public void deleteElement(UUID elementUUid, HttpStatus expextedStatus) {
```
[the remark this answers is not in the document]
Comment (JacquesBorsenberger) [reply src/test/java/org/gridsuite/explore/server/ExploreTest.java]: removed

**Review by etiennehomer — APPROVED** [review]
Tests OK now

File: src/main/java/org/gridsuite/explore/server/services/IDirectoryElementsService.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/explore/server/services/IDirectoryElementsService.java:26]: ok

File: src/main/java/org/gridsuite/explore/server/services/ExploreService.java [hunk 15]
```diff
@@ -4,49 +4,57 @@
...
 
         this.directoryService = directoryService;
         this.studyService = studyService;
         this.contingencyListService = contingencyListService;
         this.filterService = filterService;
+        this.genericServices = Map.of(
+            FILTER, filterService,
+            CONTINGENCY_LIST, contingencyListService,
+            STUDY, studyService,
+            DIRECTORY, this);
```
[the remark this answers is not in the document]
Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/explore/server/services/ExploreService.java:57]: then directory service should have ExploreService as member (to call deleteElement) 
don't know which is better

**Review by SlimaneAmar — APPROVED** [review]

**Review by etiennehomer — APPROVED** [review]

**Pull request merged**